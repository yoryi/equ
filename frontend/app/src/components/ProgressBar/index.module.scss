:local(.container) {
    display: flex;
    flex-direction: column;

    :local(.labels) {
        position: relative;
        display: flex;
        min-height: 20px;
        margin-bottom: 12px;

        :local(.value) {
            position: absolute;
        }

        &:last-child {
            margin-top: 8px;
            margin-bottom: 0;

            :local(.value) {
                transform: translateX(50%);
            }
        }

        &:not(:last-child) :local(.value):first-child:last-child {
            transform: translateX(-50%);
        }
    }

    :local(.bar) {
        position: relative;
        width: 100%;
        height: 16px;
        background-color: var(--light-blue);
        border-radius: 12px;

        :local(.value) {
            height: 16px;
            background-color: var(--main-blue);
            border-radius: 12px;
        }

        :local(.line) {
            position: absolute;
            width: 0px;
            height: 100%;
            border-left: 1px dashed var(--light-blue);
        }
    }
}
