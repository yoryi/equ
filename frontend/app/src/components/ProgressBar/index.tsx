//Utils
import cx from "classnames"
import * as React from "react"

//Styles
import Styles from "./index.module.scss"

interface ProgressBarProps {
  title: string
  value: number | [number, number | null, number]
  max: number
  percentage?: boolean
  className?: string
}

/**
 * Renders a progress bar.
 * @param {string} title - Progress bar title
 * @param {number | [number, number | null, number]} value - Progress bar value
 * @param {number} max - Maximum value
 * @param {boolean} percentage - Set to true if the value should be displayed as percentage
 * @param {string} className - Class name applied to the component
 */
const ProgressBar: React.FC<ProgressBarProps> = ({
  title,
  value,
  max,
  percentage = false,
  className,
}) => (
  <div className={cx(Styles.container, className)}>
    {!percentage && <span className={`text-2`}>{title}</span>}

    <div
      className={cx(
        {
          [`align-items-end`]:
            (percentage && typeof value === `number`) ||
            (!!value && typeof value === `number`) ||
            (Array.isArray(value) && !value[1]),
          [`align-items-center`]: Array.isArray(value) && !!value[1],
        },
        `d-flex`,
      )}
    >
      <div className={`flex-grow-1`}>
        <div
          className={cx(Styles.labels, {
            [`justify-content-between`]: percentage,
          })}
        >
          {percentage && <span className={`text-2`}>{title}</span>}

          <div
            className={cx({
              [`position-relative flex-grow-1 px-1`]: !percentage,
            })}
          >
            {!percentage && !!value && typeof value === `number` && (
              <h5
                className={Styles.value}
                style={{ left: `${(value / max) * 100}%` }}
              >
                {value}
              </h5>
            )}

            {!percentage && Array.isArray(value) && (
              <>
                <h5
                  className={cx(Styles.value, Styles.left)}
                  style={{ right: `${100 - (value[0] / max) * 100}%` }}
                >
                  {value[0]}
                </h5>

                <h5
                  className={Styles.value}
                  style={{ left: `${(value[2] / max) * 100}%` }}
                >
                  {value[2]}
                </h5>
              </>
            )}

            {percentage && typeof value === `number` && (
              <h5>{`${
                value * 100 >= 1 ? (value * 100).toFixed(0) : `<1`
              }%`}</h5>
            )}
          </div>
        </div>

        <div className={Styles.bar}>
          {typeof value === `number` && value / max >= 0.01 && (
            <div
              className={Styles.value}
              style={{ width: `${(value / max) * 100}%` }}
            />
          )}

          {Array.isArray(value) && (
            <>
              <div
                className={cx(Styles.value, `position-absolute`)}
                style={{
                  width: `${((value[2] - value[0]) / max) * 100}%`,
                  left: `${(value[0] / max) * 100}%`,
                }}
              />

              {!!value[1] && (
                <div
                  className={Styles.line}
                  style={{ left: `${(value[1] / max) * 100}%` }}
                />
              )}
            </>
          )}
        </div>

        {Array.isArray(value) && !!value[1] && (
          <div className={Styles.labels}>
            <h5
              className={Styles.value}
              style={{ right: `${100 - (value[1] / max) * 100}%` }}
            >
              {value[1]}
            </h5>
          </div>
        )}
      </div>

      {!percentage && <span className={`ml-3 text-4`}>{max}</span>}
    </div>
  </div>
)

export default ProgressBar
