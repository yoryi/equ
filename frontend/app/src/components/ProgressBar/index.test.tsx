//Testing
import { render } from "@testing-library/react"
import faker from "faker"
import * as React from "react"

//Components
import ProgressBar from "./"

describe(`ProgressBar component`, () => {
  faker.seed(1000)

  describe(`should render and match the snapshot`, () => {
    describe(`range`, () => {
      it(`with middle value`, () => {
        const { container } = render(
          <ProgressBar
            title={faker.lorem.words(2)}
            value={[800, 900, 1000]}
            max={1000}
          />,
        )

        expect(container.firstChild).toMatchSnapshot()
      })

      it(`without middle value`, () => {
        const { container } = render(
          <ProgressBar
            title={faker.lorem.words(2)}
            value={[800, null, 1000]}
            max={1000}
          />,
        )

        expect(container.firstChild).toMatchSnapshot()
      })
    })

    it(`value`, () => {
      const { container } = render(
        <ProgressBar title={faker.lorem.words(2)} value={800} max={1000} />,
      )

      expect(container.firstChild).toMatchSnapshot()
    })

    describe(`percentage`, () => {
      it(`more than or equals 1%`, () => {
        const { container } = render(
          <ProgressBar
            title={faker.lorem.words(2)}
            value={800}
            max={1000}
            percentage
          />,
        )

        expect(container.firstChild).toMatchSnapshot()
      })

      it(`less than 1%`, () => {
        const { container } = render(
          <ProgressBar
            title={faker.lorem.words(2)}
            value={0.009}
            max={100}
            percentage
          />,
        )

        expect(container.firstChild).toMatchSnapshot()
      })
    })
  })
})
