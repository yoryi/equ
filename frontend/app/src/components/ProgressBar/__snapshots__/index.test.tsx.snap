// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar component should render and match the snapshot percentage less than 1% 1`] = `
<div
  class="container"
>
  <div
    class="align-items-end d-flex"
  >
    <div
      class="flex-grow-1"
    >
      <div
        class="labels justify-content-between"
      >
        <span
          class="text-2"
        >
          officiis pariatur
        </span>
        <div
          class=""
        >
          <h5>
            &lt;1%
          </h5>
        </div>
      </div>
      <div
        class="bar"
      />
    </div>
  </div>
</div>
`;

exports[`ProgressBar component should render and match the snapshot percentage more than or equals 1% 1`] = `
<div
  class="container"
>
  <div
    class="align-items-end d-flex"
  >
    <div
      class="flex-grow-1"
    >
      <div
        class="labels justify-content-between"
      >
        <span
          class="text-2"
        >
          atque ea
        </span>
        <div
          class=""
        >
          <h5>
            80000%
          </h5>
        </div>
      </div>
      <div
        class="bar"
      >
        <div
          class="value"
          style="width: 80%;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ProgressBar component should render and match the snapshot range with middle value 1`] = `
<div
  class="container"
>
  <span
    class="text-2"
  >
    et quia
  </span>
  <div
    class="align-items-center d-flex"
  >
    <div
      class="flex-grow-1"
    >
      <div
        class="labels"
      >
        <div
          class="position-relative flex-grow-1 px-1"
        >
          <h5
            class="value left"
            style="right: 20%;"
          >
            800
          </h5>
          <h5
            class="value"
            style="left: 100%;"
          >
            1000
          </h5>
        </div>
      </div>
      <div
        class="bar"
      >
        <div
          class="value position-absolute"
          style="width: 20%; left: 80%;"
        />
        <div
          class="line"
          style="left: 90%;"
        />
      </div>
      <div
        class="labels"
      >
        <h5
          class="value"
          style="right: 10%;"
        >
          900
        </h5>
      </div>
    </div>
    <span
      class="ml-3 text-4"
    >
      1000
    </span>
  </div>
</div>
`;

exports[`ProgressBar component should render and match the snapshot range without middle value 1`] = `
<div
  class="container"
>
  <span
    class="text-2"
  >
    fugit est
  </span>
  <div
    class="align-items-end d-flex"
  >
    <div
      class="flex-grow-1"
    >
      <div
        class="labels"
      >
        <div
          class="position-relative flex-grow-1 px-1"
        >
          <h5
            class="value left"
            style="right: 20%;"
          >
            800
          </h5>
          <h5
            class="value"
            style="left: 100%;"
          >
            1000
          </h5>
        </div>
      </div>
      <div
        class="bar"
      >
        <div
          class="value position-absolute"
          style="width: 20%; left: 80%;"
        />
      </div>
    </div>
    <span
      class="ml-3 text-4"
    >
      1000
    </span>
  </div>
</div>
`;

exports[`ProgressBar component should render and match the snapshot value 1`] = `
<div
  class="container"
>
  <span
    class="text-2"
  >
    hic aut
  </span>
  <div
    class="align-items-end d-flex"
  >
    <div
      class="flex-grow-1"
    >
      <div
        class="labels"
      >
        <div
          class="position-relative flex-grow-1 px-1"
        >
          <h5
            class="value"
            style="left: 80%;"
          >
            800
          </h5>
        </div>
      </div>
      <div
        class="bar"
      >
        <div
          class="value"
          style="width: 80%;"
        />
      </div>
    </div>
    <span
      class="ml-3 text-4"
    >
      1000
    </span>
  </div>
</div>
`;
